components-rs/bytes.rs

```diff
@@ -0,0 +1,227 @@
+        Cow::Owned(s) => s.into(),
+        Cow::Borrowed(_) => Bytes::from_underlying(ZendStringWrapper(zend_str)),
+    }
+}
+
+unsafe fn convert_to_bytes_string(zend_str: *mut ZendString) -> BytesString {
+    (*zend_str).refcount += 1; // Increment the reference count to prevent double free
+
+    match String::from_utf8_lossy((*zend_str).as_ref()) {
+        Cow::Owned(s) => s.into(),
+        Cow::Borrowed(_) => {
+            BytesString::from_bytes_unchecked(Bytes::from_underlying(ZendStringWrapper(zend_str)))
+        }
+    }
+}
+
+macro_rules! set_string_field {
+    ($ptr:expr, $str:expr, $field:ident) => {{
+        if $ptr.is_null() || $str.is_null() {
+            return;
+        }
+
+        let object = &mut *$ptr;
+        object.service = convert_to_bytes_string($str);
+    }};
+}
+
+// Insert an element in the given hashmap field.
+macro_rules! insert_hashmap {
+    ($ptr:expr, $key:expr, $value:expr, $field:ident) => {{
+        if $ptr.is_null() {
+            return;
+        }
+        let object = &mut *$ptr;
+        let key = convert_to_bytes_string($key);
+        object.$field.insert(key, $value);
+    }};
+}
+
+// Remove an element from the given hashmap field.
+macro_rules! remove_hashmap {
+    ($ptr:expr, $key:expr, $field:ident) => {{
+        if $ptr.is_null() {
+            return;
+        }
+        let object = &mut *$ptr;
+        let key = convert_to_bytes_string($key);
+        object.$field.remove(&key);
+    }};
+}
+
+// Check if an element exists in the given hashmap field.
+macro_rules! exists_hashmap {
+    ($ptr:expr, $key:expr, $field:ident) => {{
+        if $ptr.is_null() {
+            return false;
+        }
+        let object = &mut *$ptr;
+        let key = convert_to_bytes_string($key);
+        return object.$field.contains_key(&key);
+    }};
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_service_(ptr: *mut SpanBytes, str: *mut ZendString) {
+    set_string_field!(ptr, str, service);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_name_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, name);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_resource_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, resource);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_type_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, r#type);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_span_meta_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(ptr, key, convert_to_bytes_string(val), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_span_meta_(ptr: *mut SpanBytes, key: *mut ZendString) {
+    remove_hashmap!(ptr, key, meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_get_span_meta_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+) -> *mut c_char {
```

Comment:
We generally want to return ddcommon_ffi::CharSlice from ffi. We do not want to do any memory allocations or such inside getters.